src/db/pool.cr

```diff
@@ -38,20 +62,28 @@ module DB
     end
 
     def checkout : T
-      resource = if @available.empty?
-                   if can_increase_pool
-                     build_resource
+      res = sync do
+        resource = if @idle.empty?
+                     if can_increase_pool?
+                       @inflight += 1
+                       r = unsync { build_resource }
+                       @inflight -= 1
+                       r
+                     else
+                       unsync { wait_for_available }
```

Comment:
By the time `unsync` has re-acquired the lock, there may be no idle connections to pick. Why not use a buffered channel as the idle connection pool itself?

Reply:
Never mind, that was the next commit in the PR... :)